src/main/java/org/squiddev/cobalt/function/LuaInterpreter.java

```diff
@@ -640,6 +655,850 @@ A C R(A+3), ... ,R(A+2+C):= R(A)(R(A+1),
 		}
 	}
 
+	private static Function<UnwindableRunnable, UnwindableCallable<EvalCont>> continuation(LuaState state, int pc, Prototype proto) {
+		final DebugHandler handler = state.debug;
+		final int instr = proto.code[pc] >> POS_OP & MAX_OP;
+
+		return f -> {
+			final UnwindableCallable<EvalCont> callable = di -> {
+				final DebugState ds = DebugHandler.getDebugState(state);
+				handler.onInstruction(ds, di, di.pc);
+//				state.instructionHits[instr]++;
+				f.run(di);
+				//noinspection ReturnOfNull
+				return null;
+			};
+
+			proto.partiallyEvaluated.set(pc);
+			proto.compiledInstrs[pc] = callable;
+
+			return callable;
+		};
+	}
+
+	private static Function<UnwindableCallable<EvalCont>, UnwindableCallable<EvalCont>> rawCont(LuaState state, int pc, Prototype proto) {
+		final DebugHandler handler = state.debug;
+		final int instr = proto.code[pc] >> POS_OP & MAX_OP;
+
+		return raw -> {
+			final UnwindableCallable<EvalCont> f = di -> {
+				final DebugState ds = DebugHandler.getDebugState(state);
+				handler.onInstruction(ds, di, di.pc);
+//				state.instructionHits[instr]++;
+				return raw.call(di);
+			};
+
+			proto.partiallyEvaluated.set(pc);
+			proto.compiledInstrs[pc] = f;
+
+			return f;
+		};
+	}
+
+	static Varargs partialEval(final LuaState state, DebugFrame di, LuaInterpretedFunction function) throws LuaError, UnwindThrowable {
+		Prototype proto = function.getPrototype();
+		EvalCont cont;
+
+		while (true) {
+			if (proto.partiallyEvaluated.get(di.pc)) {
```

Comment:
Do we need the bitset - would it be sufficient to do:

```java
var x = proto.compiledInstrs[di.pc];
if (x == null) x = partialEvalStep(state, di, function);
cont = x.call(di);
```

Reply:
The bit set is actually a setup for some experimental optimisations. Sadly, my attempts to implement them have so far only resulted in poor performance.

See https://github.com/SquidDev/Cobalt/commit/24713e0a523cccdefca0e3692275a738a5fb6210#diff-e872be1e4fc4940e53d4116bf7d9fa5fR1348 for a good example of one such change. `FORLOOP` caches its `step` and `limit` values, which are guaranteed not to change (this invariant could perhaps be broken by using the `debug` API though). To "load" new values of `step` and `limit` into the `for` loop, `FORPREP` forces the re-evaluation of the `FORLOOP` lambda (https://github.com/SquidDev/Cobalt/commit/24713e0a523cccdefca0e3692275a738a5fb6210#diff-e872be1e4fc4940e53d4116bf7d9fa5fR1396). This could once again be accomplished with `null` values in `compiledInstrs` but I thought that if there are many instructions to `null` a bit set would be faster.

The need to `null` multiple instructions would arise in optimisations which would operate on the context of more than one or two instructions. For example, I'd like to try converting whole runs of arithmetic operations into lambdas. Consider
```lua
a = (a * b) % c
```
This could produce a lambda like
```java
di -> {
  try {
    double a = di.stack[a].doubleOrThrow();
    double b = di.stack[b].doubleOrThrow();
    double c = di.stack[c].doubleOrThrow();
    di.stack[a] = valueOf((a * b) % c);
  } catch (LuaTypeAssertionException e) {
    p.partiallyEvaluated.clear();
    p.forceSafeEval.set(pc);
    return new EvalCont(pc);
  }
}
```
The evaluator would first generate these unsafe optimised lambdas. Runtime type errors in such expressions would re-enter the evaluator at the same address, forcing a safe implementation to be generated via another bit set, `forceSafeEval`. That way number-crunching code with proper type assertions (or well-behaved callers) could be replaced by efficient Java implementations but correct semantics would be preserved, for the cost of at most one re-evaluation of every instruction.

As I've said, so far my attempts have been fruitless. The commit https://github.com/SquidDev/Cobalt/pull/38/commits/22ffbb00b3862ec8dcf4ae5da718ca8e6a3a9974 (for which I had high hopes) turned out to decrease performance. I wanted to take that optimisation a couple steps further by replacing the common `LOADK`'s before `FORPREP`'s with lambdas which would simply increment the PC and directly call the compiled `FORPREP` instruction.

It's very well possible that all of this could be accomplished without the `partiallyEvaluated` bit set and it's more than likely that none of the above suggestions will work. Perhaps this work is better left for another PR...